cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,883 @@
+
+cloneP :: forall curve. (BLS_P curve) => P curve -> IO (P curve)
+cloneP (P a) = do
+  b <- mallocForeignPtrBytes (sizeP (Proxy @curve))
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap (sizeP (Proxy @curve))
+  return (P b)
+
+withAffine :: forall a curve. Affine curve -> (AffinePtr curve -> IO a) -> IO a
+withAffine (Affine p) go = withForeignPtr p (go . AffinePtr)
+
+withNewAffine :: forall curve a. (BLS_P curve) => (AffinePtr curve -> IO a) -> IO (a, Affine curve)
+withNewAffine go = do
+  p <- mallocForeignPtrBytes (sizeAffine (Proxy @curve))
+  x <- withForeignPtr p (go . AffinePtr)
+  return (x, Affine p)
+
+withNewAffine_ :: (BLS_P curve) => (AffinePtr curve -> IO a) -> IO a
+withNewAffine_ = fmap fst . withNewAffine
+
+withNewAffine' :: (BLS_P curve) => (AffinePtr curve -> IO a) -> IO (Affine curve)
+withNewAffine' = fmap snd . withNewAffine
+
+
+withPT :: PT -> (PTPtr -> IO a) -> IO a
+withPT (PT pt) go = withForeignPtr pt (go . PTPtr)
+
+withNewPT :: (PTPtr -> IO a) -> IO (a, PT)
+withNewPT go = do
+  p <- mallocForeignPtrBytes sizePT
+  x <- withForeignPtr p (go . PTPtr)
+  return (x, PT p)
+
+withNewPT_ :: (PTPtr -> IO a) -> IO a
+withNewPT_ = fmap fst . withNewPT
+
+withNewPT' :: (PTPtr -> IO a) -> IO PT
+withNewPT' = fmap snd . withNewPT
+
+instance BLS_P Curve1 where
+  _sizeP _ = fromIntegral c_size_blst_p1
+  _compressedSizeP _ = 48
+  _serializedSizeP _ = 96
+  _sizeAffine _ = fromIntegral c_size_blst_affine1
+
+instance BLS_P Curve2 where
+  _sizeP _ = fromIntegral c_size_blst_p2
+  _compressedSizeP _ = 96
+  _serializedSizeP _ = 192
+  _sizeAffine _ = fromIntegral c_size_blst_affine2
+
+sizePT :: Num i => i
+sizePT = fromIntegral c_size_blst_fp12
+
+
+---- Curve operations
+
+-- | BLS curve operations. Class methods are low-level; user code will want to
+-- use higher-level wrappers such as 'add', 'mult', 'cneg', 'neg', etc.
+class BLS_Curve curve where
+  c_blst_on_curve :: PPtr curve -> IO Bool
+
+  c_blst_add :: PPtr curve -> PPtr curve -> PPtr curve -> IO ()
+  c_blst_mult :: PPtr curve -> PPtr curve -> ScalarPtr -> CInt -> IO ()
+  c_blst_cneg :: PPtr curve -> Bool -> IO ()
+
+  c_blst_hash :: PPtr curve -> Ptr CChar -> CSize -> Ptr CChar -> CSize -> Ptr CChar -> CSize -> IO ()
+  c_blst_compress :: Ptr CChar -> PPtr curve -> IO ()
+  c_blst_serialize :: Ptr CChar -> PPtr curve -> IO ()
+  c_blst_uncompress :: AffinePtr curve -> Ptr CChar -> IO CInt
+  c_blst_deserialize :: AffinePtr curve -> Ptr CChar -> IO CInt
+
+  c_blst_in_g :: PPtr curve -> IO Bool
+  c_blst_to_affine :: AffinePtr curve -> PPtr curve -> IO ()
+  c_blst_from_affine :: PPtr curve -> AffinePtr curve -> IO ()
+  c_blst_affine_in_g :: AffinePtr curve -> IO Bool
+  c_blst_generator :: PPtr curve
+
+  c_blst_p_is_equal :: PPtr curve -> PPtr curve -> IO Bool
+  c_blst_p_is_inf :: PPtr curve -> IO Bool
+
+instance BLS_Curve Curve1 where
+  c_blst_on_curve = c_blst_p1_on_curve
+
+  c_blst_add = c_blst_p1_add
```

Comment:
I think because we eventually use this function to FFI call the `add_or_double` function from the C library it makes sense to change this to.
`c_blst_add = c_blst_p1_add_or_double` or `c_blst_add_or_double = c_blst_p1_add_or_double`

And then change all instances of `c_blst_add` to `c_blst_add_or_double`

Reply:
Done.